core: Add support of non-UTF-8 encodings to the form loader

Korne127 opened the pull request:
Edit: To read the current summary of this pull request, [click here](https://github.com/ruffle-rs/ruffle/pull/13346#issuecomment-2166160151) (or scroll down).

----

The form loader is currently decoding all files as UTF-8.
Therefore, files that aren't encoded as UTF-8 are decoded wrongly if they're loaded with the form loader, and the content isn't saved correctly.
In this case, non-ASCII characters like European letters with [diacritics](https://en.wikipedia.org/wiki/Diacritic) aren't decoded correctly and therefore cannot be displayed. For non-Latin scripts, the impact is much more severe as almost everything is decoded wrongly and therefore unreadable.

This pull request fixes this. It adds the crate [chardetng](https://hsivonen.fi/chardetng/), which is used to detect the encoding. If necessary, the form data is converted into UTF-8.

kmeisthax commented:
The returned data should be accompanied by an HTTP response header listing out the encoding. Do we know if Flash uses that header?

Korne127 commented:
@kmeisthax The response header does not always include a charset. The example through which I found this bug was a .txt file that was loaded with the form loader. The file is encoded in Latin-1 but encoded (as currently everything is) in UTF-8.
There is no way to always correctly encode such a txt file with a 100% guarantee as there is no given charset. UTF-8 texts can contain byte order marks (BOMs) indicating the text is encoded in UTF-8 (but don't have to), but in general, the way to display all texts correctly is like this. Firefox itself uses the same crate when opening the same txt file (and displays a warning in the console that it's not guaranteed to be the right encoding).
This way, texts should almost always be encoded with the right encoding in Ruffle and displayed properly.

@adrian17 mentioned in the Discord server that this is likely not how Flash itself did it. Latin-1 was more widely used 10-20 years ago, there might have just been a test to differenciate between Latin-1 and UTF-8 (or also used a given charset if existing). Flash might have decoded texts in other little-used encodings wrongly.
Technically, Flash decoding such texts wrongly could have been used in a way (e.g. while storing byte data) although that's not very likely.

n0samu commented:
My hunch is that Flash Player decodes non-UTF8 text according to the system codepage. I tried converting the txt file in Ruffle's loadvariables test to Shift-JIS (and adding some Japanese characters), and Flash Player seems to try to interpret the text as Windows-1252 (which produces garbage output). I don't think smarter behavior is a bad thing though! Just need to make sure it's reliable.

kmeisthax commented:
System codepage is basically guaranteed to create mojibake, and we don't have access to it in HTML, so I'm perfectly fine with sniffing the encoding in that case. Especially if we're using Firefox's sniffer which should be well tested.

In the case where the form response *does* have valid encoding information, is it possible to use that?

Korne127 commented:
I looked into this for a while and tried to find out how Flash itself exactly determined it.

To do this, I created a test SWF file which loads several text files with different encodings and different HTTP headers.
There are twelve files, each combination out of the encoding (UTF-8, ISO Latin-1 and Shift JIS) and the charset in the HTTP header (None, UTF-8, ISO Latin-1 and Shift JIS). The loaded variables are printed onto a text box.
You can try it out here: https://korne127.de/flash_test/encoding_test/Test.swf

Surprisingly, when I tested it, only the UTF-8 is decoded correctly (both with the Flash plugin and the desktop version):
<img width="843" alt="Decoding results without HTTP header" src="https://github.com/ruffle-rs/ruffle/assets/33464213/f4165184-00f2-4bec-b1b0-f8353753638e">

And the HTTP header didn't change anything in my tests (CharISO means that the HTTP header includes ISO Latin-1 as charset):
<img width="839" alt="Decoding results with HTTP header" src="https://github.com/ruffle-rs/ruffle/assets/33464213/5d5a8da9-7a4e-4fce-a75b-46788cae27b9">

As the ISO Latin-1 text file loaded by the SWF I found is correctly decoded by Flash, I'll assume that there is some other factor which determines the encoding Flash uses.

@kmeisthax @n0samu Can you test this as well since this might behave differently on different platforms?
<br>

Independently to what Flash does, I think it might make sense to use this PR's approach and correctly determine the charset.  
Theoretically, someone could have stored binary data in a text file that is then read by Flash as random characters that are used in a specific way, in that case determining a charset would change the behaviour, but I find that rather unlikely.

I'll look into using the HTTP header if one exists instead of `chardetng` (this obviously only works for remote files though).

n0samu commented:
This is the result I got opening the SWF via its URL in the Flash projector on Windows 10. Looks like the same result as yours, hope it helps. Do you have an example of a game that this PR helps with btw? Just wondering
```
UTF-8.txt
var2: ÄÖÜ ß
var1: Example Text in UTF8:

Iso Latin-1.txt
var2: Ŗڠ
var1: Example Text in Iso Latin-1:

Shift JIS.txt
var2: Љ¼¼
var1: Example Text in Shift JIS:

CharUTF UTF-8.txt
var2: ÄÖÜ ß
var1: Example Text in UTF8:

CharUTF Iso Latin-1.txt
var2: Ŗڠ
var1: Example Text in Iso Latin-1:

CharUTF Shift JIS.txt
var2: Љ¼¼
var1: Example Text in Shift JIS:

CharISO UTF-8.txt
var2: ÄÖÜ ß
var1: Example Text in UTF8:

CharISO Iso Latin-1.txt
var2: Ŗڠ
var1: Example Text in Iso Latin-1:

CharISO Shift JIS.txt
var2: Љ¼¼
var1: Example Text in Shift JIS:

CharShift UTF-8.txt
var2: ÄÖÜ ß
var1: Example Text in UTF8:

CharShift Iso Latin-1.txt
var2: Ŗڠ
var1: Example Text in Iso Latin-1:

CharShift Shift JIS.txt
var2: Љ¼¼
var1: Example Text in Shift JIS:
```

Korne127 commented:
Thank you! Yeah, it looks like the same result, and I got the same with the plugin version as well. That's quite interesting, I wonder what's causing Flash to only decode the UTF-8 files correctly in this test (as Flash is correctly decoding ISO Latin-1 files in other SWFs).

And yeah, I have a game; I noticed this issue originally in [Paraplüsch](parapluesch.de/whiskystore/test.htm). If you choose the German flag, the "Vorgeschichte Überspringen" (Skip Intro in German) turns into "Vorgeschichte berspringen" as soon as the [text file](parapluesch.de/whiskystore/deut.txt) is loaded whose content overwrites that text field. Text from that file is seen in other places in the game as well, where all letters with diacritics are not displayed.  
That happens because the text is encoded as ISO Latin-1 and read as UTF-8 by Ruffle. This is not the case in Flash which displays all texts correctly.

https://github.com/ruffle-rs/ruffle/assets/33464213/aee763d0-c188-4423-ab62-5bfe8f55ab91



I also implemented kmeisthax's idea to use the HTTP header encoding and only `chardetng` if the HTTP header doesn't include one. This worked great on desktop, I'll test the web version tomorrow and then update this pull request.

n0samu commented:
This game is SWFv5, which stored strings with the system encoding instead of UTF-8, so it makes sense that external strings loaded from SWFv5 would behave the same way.

Korne127 commented:
I committed the idea to use the response header. With the current commits, the encoding specified in the response header is used if one is existing, otherwise `chardetng`.

@n0samu Thank you for the tip! I tested it again with different SWF versions and can now confirm that the behaviour is the same in SWF version 6, 8, 10 and 43. I haven't tested version 5 yet as my test SWF uses some functions that only exist in SWF versions 6 and upwards.

Why do you say that the game is SWFv5 though? I downloaded `anstalt_kore.swf`, and according to FFDec, the SWF version is 8.

I think an approach like this could also be part of the solution for #8390. At least if we don't have access to the system codepage and / or it's not reliable for some reason, detecting the encoding through `chardetng` seems like a reasonable alternative.

n0samu commented:
@Korne127 Oh, I'd only looked at `offtext_deut.swf`, which I assumed was loading the text file. Seems like I was wrong though, since the loadVariables call indeed comes from `anstalt_kore.swf`, which is SWFv8. Sorry about that

Korne127 commented:
Hey :)
I spent some time figuring out what's different about Paraplüsch that it loads text files in Iso Latin-1 (or to be exact Windows-1252) while all my test SWF files used UTF-8.
And after a while, I found it: It contains this command in the first frame: `System.useCodepage = true;`
As @n0samu assumed, Flash loads the text files in the system codepage, but only if this variable has been set to `true`.
Otherwise, UTF-8 is used if the SWF version is 6 or higher. If the SWF version is 5 or lower, it uses Windows-1252 on Windows and a slightly different custom encoding on macOS (the letter with diacritics have the same encoding as in Windows-1252 but some other symbols don't).

I updated this pull request accordingly to my findings.
Ruffle doesn't always have access to the system codepage and it's not reliably the correct encoding. Therefore, if `System.useCodepage` has been set to `true`, Ruffle now uses the encoding in the HTTP response header if existing, and guesses it via `chardetng` otherwise.
If `System.useCodepage` is `false`, Ruffle uses UTF-8 if the SWF version is 6 or higher and Windows-1252 otherwise.
This way, Ruffle is able to load non-UTF-8 files and display their characters correctly.

I also added some tests to make sure that there won't be regressions that cause Ruffle to load files incorrectly again.

adrian17 commented on core/src/backend/navigator.rs:
```diff
@@ -682,3 +690,14 @@ pub fn fetch_path<NavigatorType: NavigatorBackend>(
+/// Parses and returns the encoding out of an HTTP header content type string
+/// if existing.
+pub fn get_encoding(content_type: String) -> Option<&'static Encoding> {
+    let content_vector = content_type.split("charset=").collect::<Vec<&str>>();
```
I think you can call `split_once` and skip a Vec this way?

Korne127 replied:
You're right; thanks! :)

adrian17 commented on core/src/backend/navigator.rs:
```diff
@@ -682,3 +690,14 @@ pub fn fetch_path<NavigatorType: NavigatorBackend>(
+
+/// Parses and returns the encoding out of an HTTP header content type string
+/// if existing.
+pub fn get_encoding(content_type: String) -> Option<&'static Encoding> {
```
should be possible to pass `&str` without converting `.to_string()` here, I think

Korne127 replied:
Yeah, that's true! Thank you :)

adrian17 commented on core/src/avm1/globals/system.rs:
```diff
@@ -265,8 +265,8 @@ pub struct SystemProperties {
-    /// If true then the system codepage should be used instead of unicode for text files
-    /// If false then unicode should be used
+    /// If true, the system codepage should be used for text files
+    /// If false, UTF-8 should be used for SWF version >= 6 and ISO Latin-1 for SWF version <= 5
```
to be sure, the behavior on <=5 isn't the same as-if `useCodepage` was set to `true`?

Korne127 replied:
Yeah. I tested it on my macOS system and in a Windows VM with the codepage both set to German (Windows-1252) and Japanese (Shift-JIS).
When useCodepage has been set to true, the local codepage has always been used. But when I tried an SWFv5 file, it has always been encoded as Windows-1252 (to be exact, a custom but similar encoding on macOS), even when the codepage was Japanese.

In this image you can see the same code loading some texts in different encodings on a Windows system with a Japanese codepage. The only difference is
SWFv10 / SWFv10 with useCodepage true / SWFv5
<img width="1023" alt="Bildschirmfoto 2024-06-19 um 01 52 02" src="https://github.com/ruffle-rs/ruffle/assets/33464213/dfb4a49f-0453-441f-a1f0-e1ac7f55b76a">

adrian17 commented on frontend-utils/src/backends/navigator.rs:
```diff
@@ -271,7 +272,16 @@ impl<F: FutureSpawner, I: NavigatorInterface> NavigatorBackend for ExternalNavig
+                let text_encoding =
+                    if let Some(content_type) = response.headers().get("Content-Type") {
+                        if let Ok(content_type_str) = content_type.to_str() {
+                            get_encoding(content_type_str)
```
would this work? To reduce the nesting
```rs
let text_encoding = response
    .headers()
    .get("Content-Type")
    .and_then(|header| header.to_str())
    .and_then(|header| get_encoding(content_type_str));
```

Korne127 replied:
Sure, I can do that

Korne127 replied:
That approach works if you add a `.ok()` after the `.to_str()`
Thanks for the suggestion :)

adrian17 commented on tests/framework/src/backends/navigator.rs:
```diff
@@ -51,6 +46,16 @@ impl SuccessResponse for TestResponse {
         self.redirected
     }
 
+    fn next_chunk(&mut self) -> OwnedFuture<Option<Vec<u8>>, Error> {
```
Can you please drag the code back to original place, before `fn status`? :D

Korne127 replied:
Why? This way, the implementation order is the same as the trait order (and the implementation order in every other implementation). It's cleaner to have it consistent (and shows up as warning otherwise in some IDEs).

kjarosh replied:
The disadvantage of moving code around is that its blame history is lost (it's harder to track changes). IIRC there are more places where the order is not preserved.

So I guess it's a decision of what's better in this case (preserving the order vs preserving blame history).

Maybe it's also worth considering adding a lint for that when we decide that preserving the order is important.

Korne127 replied:
Hmm, I see. I get that point, but personally, I think code is slightly changed all the time, especially in a project that moves as fast as Ruffle. When you use blame, you often need to look at the changes *before* the last one as refactors and other minor changes are frequent, so changing the order to improve code quality doesn't disrupt workflows.
There can be reasons to purposely have a different order in some places, but I think it's clean to generally try to preserve it.

kjarosh commented on tests/tests/swfs/avm1/form_loader_encoding_1/test.as:
```diff
@@ -0,0 +1,18 @@
+ * This test is the first test out of a series of tests, testing the encoding the form loader uses to decode
+ * text files.
+ * This test tests whether a UTF-8 text file loaded in an SWF file with SWF version 6 is decoded as UTF-8.
+ * Flash decodes invalid UTF-8 characters (such as some characters encoded in ISO Latin-1) differently, and
+ * as of June 2024, Ruffle doesn't implement this. Therefore, this test only tests a UTF-8 file.
```
I would suggest adding those files anyway, commenting out the code that verifies their values, and adding a TODO. Alternatively a 4th version of this test that is a known failure?

Korne127 replied:
I think adding a new test just for this is a bit overkill.
Essentially, this means that when Flash loads a file with invalid UTF-8 characters, it decodes them as some seemingly random symbols while Ruffle just only displays the valid parts.
Realistically, that behaviour isn't really wanted (if you want any symbols, you can just put them as UTF-8 in a file, this mostly happens when a file is wrongly encoded).

I can add an explicit TODO as you suggested. But I'm not completely sure if we want to implement this; it's likely not very easy. I can imagine that we'd need to write an own UTF-8 parser, and going through every text file like this would also include a slight (although probably insignificant) performance penalty. Because of this, I just wrote it as a notice.

kjarosh replied:
The truth is that every observable behavior may be depended upon. There are some obfuscated SWFs that rely on decoding invalid UTF-8 (e.g. https://github.com/ruffle-rs/ruffle/issues/15919), or some SWFs currently panic in Ruffle due to decoding errors (e.g. https://github.com/ruffle-rs/ruffle/pull/12221, I myself have encountered some).

I was thinking in terms of:
1. some SWFs may depend on this behavior,
2. this behavior may help resolve similar issues related to decoding invalid UTF-8.

TBH I'm not sure how exactly they are related, just noticed that all of them have decoding invalid UTF-8 in common :)

Korne127 replied:
Okay, that's actually pretty reasonable. Thanks for the explanation.
What do you think would be the best approach?

kjarosh replied:
I would personally suggest adding a 4th test that is a known failure and describing the issue there, but a TODO comment with commented out code is also okay if you don't want to add another test :)

If you have any observations etc. related to FP in this case it would also be great to include them in comments!

Korne127 replied:
Done :)
